modules/infra/datapath/gr_trace.h

```diff
@@ -5,9 +5,23 @@
 #ifndef _GR_INFRA_PACKET_TRACE
 #define _GR_INFRA_PACKET_TRACE
 
+#include <rte_graph.h>
 #include <rte_mbuf.h>
 
+#include <sys/queue.h>
+
 // Write a log message with detailed packet information.
 void trace_log_packet(const struct rte_mbuf *m, const char *node, const char *iface);
 
+// Callback associated with each node that will be invoked by gr_trace_dump
+// to format each individual trace items.
+typedef int (*gr_trace_format_cb_t)(char *buf, size_t buf_len, const void *data, uint16_t data_len);
+
+// Format the buffered trace items and empty the buffer.
+// Return the number of bytes written to buffer or a negative value on error.
+int gr_trace_dump(char *buf, size_t buf_len);
```

Comment:
It is hard to see when an error would be detected in the current implementation.